File: src/jjoller/foxinabox/RandomDealer.java
package jjoller.foxinabox;

import java.util.*;

/**
 * Is responsible for dealing the cards. This implementation deals random cards.
 */
public class RandomDealer implements Dealer {

    public RandomDealer() {
        deck = new ArrayList<>(Card.values().length);
        deck.addAll(Arrays.asList(Card.values()));
        removed = new ArrayList<>(25);
        random = new Random();
    }


    private final List<Card> deck;
    private final List<Card> removed;
    private final Random random;

    public void reset() {
        deck.addAll(removed);
        removed.clear();
    }

    public void dealTableCards(TexasHand hand) {

        switch (hand.phase()) {
            case PRE_FLOP:
                EnumSet<Card> flop = EnumSet.of(removeRandomFromDeck(),
                        removeRandomFromDeck(), removeRandomFromDeck());
                hand.setFlop(flop);
                break;
            case FLOP:
                hand.setTurn(removeRandomFromDeck());
                break;
            case TURN:
                hand.setRiver(removeRandomFromDeck());
                break;
            default:
                throw new IllegalStateException();
        }
    }

    public void dealHoleCards(TexasHand.Player player) {

        player.setCards(EnumSet.of(removeRandomFromDeck(),
                removeRandomFromDeck()));
    }

    private Card removeRandomFromDeck() {
        Card card = deck.remove(random.nextInt(deck.size()));
        removed.add(card);
        return card;
    }

    public int smallBlind() {
        return 1;
    }

    public int bigBlind() {
        return 2;
    }

}

File: src/jjoller/foxinabox/RandomDealer.kt
package jjoller.foxinabox

import java.util.*

/**
 * Is responsible for dealing the cards. This implementation deals random cards.
 */
class RandomDealer : Dealer {

    private val deck: MutableList<Card>
    private val removed: MutableList<Card>
    private val random: Random

    init {
        deck = ArrayList<Card>(Card.values().size)
        deck.addAll(Arrays.asList(*Card.values()))
        removed = ArrayList<Card>(25)
        random = Random()
    }

    override fun reset() {
        deck.addAll(removed)
        removed.clear()
    }

    override fun dealTableCards(hand: TexasHand) {

        when (hand.phase()) {
            Phase.PRE_FLOP -> {
                hand.flop = EnumSet.of(removeRandomFromDeck(),
                        removeRandomFromDeck(), removeRandomFromDeck())
            }
            Phase.FLOP -> hand.turn = removeRandomFromDeck()
            Phase.TURN -> hand.river = removeRandomFromDeck()
            else -> throw IllegalStateException()
        }
    }

    override fun dealHoleCards(player: TexasHand.Player) {

        player.setCards(EnumSet.of(removeRandomFromDeck(),
                removeRandomFromDeck()))
    }

    private fun removeRandomFromDeck(): Card {
        val card = deck.removeAt(random.nextInt(deck.size))
        removed.add(card)
        return card
    }

    override fun smallBlind(): Int {
        return 1
    }

    override fun bigBlind(): Int {
        return 2
    }

}

